File: java-source/src/main/java/com/iou/flow/IOUTransferFlow.java
package com.iou.flow;

import co.paralleluniverse.fibers.Suspendable;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableSet;
import com.iou.contract.IOUContract;
import com.iou.state.IOUState;
import net.corda.core.contracts.*;
import net.corda.core.crypto.CompositeKey;
import net.corda.core.crypto.DigitalSignature;
import net.corda.core.crypto.Party;
import net.corda.core.flows.FlowException;
import net.corda.core.flows.FlowLogic;
import net.corda.core.transactions.SignedTransaction;
import net.corda.core.transactions.TransactionBuilder;
import net.corda.flows.FinalityFlow;

import java.io.FileNotFoundException;
import java.security.KeyPair;
import java.util.List;
import java.util.Set;

import static kotlin.collections.CollectionsKt.single;

public class IOUTransferFlow {
    public static class Initiator extends FlowLogic<SignedTransaction> {
        private final StateRef stateRef;
        private final Party newRecipient;

        public Initiator(StateRef stateRef, Party newRecipient) {
            this.stateRef = stateRef;
            this.newRecipient = newRecipient;
        }

        @Suspendable
        public SignedTransaction call() throws FlowException {
            // Stage 1 - Retrieving the existing IOU from the vault.
            final TransactionState inputState = getServiceHub().getVaultService().statesForRefs(ImmutableList.of(stateRef)).get(stateRef);
            final IOUState inputStateData = (IOUState) inputState.getData();

            // Stage 2 - This flow-logic can only be initiated by the current recipient.
            if (!getServiceHub().getMyInfo().getLegalIdentity().getName().equals(inputStateData.getRecipient().getName())) {
                throw new IllegalArgumentException("IOU transfer can only be initiated by the IOU's current recipient");
            }

            // Stage 3 - Creating the output state.
            final Party sender = inputStateData.getSender();
            final IOUState outputState = new IOUState(
                    inputStateData.getIOUValue(),
                    sender,
                    newRecipient,
                    new IOUContract(),
                    inputStateData.getLinearId()
            );

            // Stage 4 - Creating the transfer command.
            final List<CompositeKey> participants = inputState.getData().getParticipants();
            participants.add(newRecipient.getOwningKey());
            final Command txCommand = new Command(new IOUContract.Transfer(), participants);

            // Stage 5 - Generating an unsigned transaction.
            final Party notary = single(getServiceHub().getNetworkMapCache().getNotaryNodes()).getNotaryIdentity();
            final StateAndRef inputStateAndRef = new StateAndRef(inputState, stateRef);
            final TransactionBuilder unsignedTx = new TransactionType.General.Builder(notary).withItems(inputStateAndRef, outputState, txCommand);

            // Stage 6 - Verifying the transaction.
            try {
                unsignedTx.toWireTransaction().toLedgerTransaction(getServiceHub()).verify();
            } catch (FileNotFoundException ex) {
                throw new RuntimeException("Required attachment was not found in storage.", ex);
            } catch (TransactionResolutionException ex) {
                throw new RuntimeException("An input points to a transaction not found in storage.", ex);
            }

            // Stage 7 - Signing the transaction.
            final KeyPair keyPair = getServiceHub().getLegalIdentityKey();
            final SignedTransaction partSignedTx = unsignedTx.signWith(keyPair).toSignedTransaction(false);

            // Stage 8 - Collecting the counterparty signatures.
            final DigitalSignature.WithKey extraSig1 = subFlow(new IOUTransferSubflow.Initiator(partSignedTx, sender));
            final DigitalSignature.WithKey extraSig2 = subFlow(new IOUTransferSubflow.Initiator(partSignedTx, newRecipient));
            final SignedTransaction fullySignedTx = partSignedTx.plus(extraSig2).plus(extraSig1);

            // Stage 9 - Calling FinalityFlow.
            final Set<Party> parties = ImmutableSet.of(sender, inputStateData.getRecipient(), newRecipient);
            subFlow(new FinalityFlow(fullySignedTx, parties));

            return fullySignedTx;
        }
    }
}

File: kotlin-source/src/main/kotlin/com/iou/kotlin/flow/IOUTransferFlow.kt
package com.iou.kotlin.flow

import co.paralleluniverse.fibers.Suspendable
import com.iou.kotlin.contract.IOUContract
import com.iou.kotlin.state.IOUState
import net.corda.core.contracts.Command
import net.corda.core.contracts.StateAndRef
import net.corda.core.contracts.StateRef
import net.corda.core.contracts.TransactionType
import net.corda.core.crypto.Party
import net.corda.core.flows.FlowLogic
import net.corda.core.transactions.SignedTransaction
import net.corda.flows.FinalityFlow

object IOUTransferFlow {
    class Initiator(val stateRef: StateRef, val newRecipient: Party) : FlowLogic<SignedTransaction>() {

        @Suspendable
        override fun call(): SignedTransaction {
            // Stage 1 - Retrieving the existing IOU from the vault.
            val inputState = serviceHub.vaultService.statesForRefs(listOf(stateRef))[stateRef]!!
            val inputStateData = inputState.data as IOUState

            // Stage 2 - This flow-logic can only be initiated by the current recipient.
            if (serviceHub.myInfo.legalIdentity.name != inputStateData.recipient.name) {
                throw IllegalArgumentException("IOU transfer can only be initiated by the IOU's current recipient")
            }

            // Stage 3 - Creating the output state.
            val outputState = inputStateData.copy(recipient = newRecipient)

            // Stage 4 - Creating the transfer command.
            val participants = inputState.data.participants + newRecipient.owningKey
            val txCommand = Command(IOUContract.Commands.Transfer(), participants)

            // Stage 5 - Generating an unsigned transaction.
            val notary = serviceHub.networkMapCache.notaryNodes.single().notaryIdentity
            val inputStateAndRef = StateAndRef(inputState, stateRef)
            val unsignedTx = TransactionType.General.Builder(notary).withItems(inputStateAndRef, outputState, txCommand)

            // Stage 6 - Verifying the transaction.
            unsignedTx.toWireTransaction().toLedgerTransaction(serviceHub).verify()

            // Stage 7 - Signing the transaction.
            val keyPair = serviceHub.legalIdentityKey
            val partSignedTx = unsignedTx.signWith(keyPair).toSignedTransaction(false)

            // Stage 8 - Collecting the counterparty signatures.
            val extraSig1 = subFlow(IOUTransferSubflow.Initiator(partSignedTx, inputStateData.sender))
            val extraSig2 = subFlow(IOUTransferSubflow.Initiator(partSignedTx, newRecipient))
            val fullySignedTx = partSignedTx + extraSig2 + extraSig1

            // Stage 9 - Calling FinalityFlow.
            val parties = setOf(inputStateData.sender, inputStateData.recipient, newRecipient)
            subFlow(FinalityFlow(fullySignedTx, parties))

            return fullySignedTx
        }
    }
}